tempesta_fw/http.c

```diff
@@ -2930,6 +2934,37 @@ tfw_http_req_process(TfwConn *conn, const TfwFsmData *data)
 		skb = NULL;
 	}
 
+	/*
+	 * If a client work through a forward proxy, then a proxy can pass it's
+	 * IP address by the first value in X-Forwarded-For
+	 */
+	s_xff = req->h_tbl->tbl[TFW_HTTP_HDR_X_FORWARDED_FOR];
+	s_ip = tfw_str_next_str_val(&s_xff);
+	nchunks = 0;
+	TFW_STR_FOR_EACH_CHUNK(c, &s_ip, end) {
+		if (!(c->flags & TFW_STR_VALUE))
+			break;
+		nchunks++;
+	}
+	s_ip.nchunks = nchunks;
+
+	if (TFW_STR_EMPTY(&s_ip) || tfw_addr_pton(&s_ip, &addr) != 0) {
+		ss_getpeername(conn->sk, &addr);
+	}
+
+	tfw_http_msg_clnthdr_val(&req->h_tbl->tbl[TFW_HTTP_HDR_USER_AGENT],
+				 TFW_HTTP_HDR_USER_AGENT, &s_user_agent);
+
+	if (!TFW_STR_EMPTY(&s_ip) || !TFW_STR_EMPTY(&s_user_agent)) {
+		TfwClient *cli = tfw_client_obtain(addr, &s_user_agent, NULL);
+		TfwClient *conn_cli = (TfwClient *)conn->peer;
+		if (cli && conn_cli != cli) {
+			tfw_connection_unlink_from_peer(conn);
+			tfw_client_put(conn_cli);
+			tfw_connection_link_peer(conn, (TfwPeer *)cli);
```

Comment:
A very subtle place. Here we're under the socket lock acquired by the TCP caller and asserted in `ss_tcp_data_ready()`, so we have no concurrent ingress requests. However, what if some other parallel softirq is processing the client and connection, e.g.on processing server response? Do we have such places which work with `TfwClient` on server responses? A comment about the operations safety is required. We had plenty of bugs on races with the connections linkage.

Reply:
I against the change. We mustn't update the `TfwCliConn->peer` on processing a random response. I suppose that a new member `TfwHttpReq->peer` is required:
- The peer in `TfwCliConn->peer` - is hop-by-hop peer, which can be an intermediate proxy.
- The peer in `TfwHttpReq->peer` - is end-to-end peer, which is (mostly but not always) browser.

Take a look at `http_resp_code_block` Frang directive. The directive adds a _server connection_ hook, and it's targeted to block clients that e.g. tries to bruteforce passwords. It was discussed in the chat some day, that we don't want to block the whole intermediate proxy, only some clients behind it. So we'll need to some how find an end-to-end peer, having only `TfwHttpReq` in sight.

With the changes proposed, `TfwHttpReq->conn->peer` will return a random end-to-end client and ban it on.